mm/ubsan/ubsan.c

```diff
@@ -160,7 +160,7 @@ static int64_t get_signed_val(FAR struct type_descriptor *type,
       unsigned extra_bits = sizeof(int64_t) * 8 - type_bit_width(type);
       uintptr_t ulong_val = (uintptr_t)val;
 
-      return ((int64_t)ulong_val) << extra_bits >> extra_bits;
+      return (int64_t)(ulong_val << extra_bits >> extra_bits);
```

Comment:
I'm a bit confused with this change. The signed vs unsigned shift has essential difference in populating a sign bit, so `0xffffffff << 1 >> 1` (for 32 bits) will be `0x7fffffff` for unsigned and `0xffffffff` for signed. What `get_signed_val` should return in case if `is_inline_int(type)` is evaluated to `true`? For 64 bits target it is obvious since `uintptr_t` has the same size as `int64_t` so `return (int64_t)(ulong_val << extra_bits >> extra_bits);` is equivalent to `*(FAR int64_t *)val`, but if the value is `int32_t` then this change will break the logic because sign bit will not be populated and `int32_t` `-1` will become `4294967295` `int64_t` that seems to be completely wrong.
Maybe it is better to keep code here as is and change `is_inline_int` to `return bits < inline_bits;` and not equal, so equal condition will be evaluated by `return *(FAR int64_t *)val;` in `get_signed_val`?

Reply:
> Maybe it is better to keep code here as is and change is_inline_int to return bits < inline_bits; and not equal, so equal condition will be evaluated by return *(FAR int64_t *)val; in get_signed_val?

I don't see how this is going to help. In my case the `type_bit_width()` is 32 so the same code will be executed nevertheless.

But I don't understand, why do we need to clear the "extra_bits" in the first place?

Here are some real values:  
Actual value is -1.  
extra_bits = 32  
val, ulong_val = 0xffffffffffffffff  

So, this should return 0xffffffffffffffff as an int64_t, right?  
Why clear any bits?

Reply:
I agree that right shift of negative values is undefined by C standard, but most compilers implement it to be compatible with integer division by 2

Reply:
Why are you saying that if value is `-1` then `ulong_val` is `0xffffffffffffffff`? Are you testing on 32 bit or 64 bit platform?

Reply:
maybe we can make something like
```
unsigned extra_bits = sizeof(int64_t) * 8 - type_bit_width(type);
uint64_t mask = 1llu << extra_bits;
uint64_t ret = (uint64_t)val & mask;

return (int64_t)(ret & (1ll << (extra_bits - 1)) != 0 ? ret | ~mask : ret);
```
? Or something similar

Reply:
https://docs.oracle.com/cd/E19120-01/open.solaris/816-5138/convert-2/index.html -- yeah, things can be even more tricky if signed extension is done implicitly. Let's try to avoid bit shift of value

Reply:
> I agree that right shift of negative values is undefined by C standard, but most compilers implement it to be compatible with integer division by 2

I agree that most systems behave in a similar fashion, but *technically* it is still UB.  
It is not a horrible bug in itself, but as UBSan calls itself it causes infinite recursion, ending up in a system crash.

> Why are you saying that if value is -1 then ulong_val is 0xffffffffffffffff? Are you testing on 32 bit or 64 bit platform?
I mean that for 64 bit platform it should be true, but not for 32 bit platform

Yes I am running it on a 64-bit system, using the simulator.  
Unfortunately, due to the overhead of UBSan and of the test code (that triggers this UB), I cannot run this on actual hardware.

I am having a look on your solution right now.